plugins/org.eclipse.fordiac.ide.typemanagement/src/org/eclipse/fordiac/ide/typemanagement/refactoring/UpdateFBInstanceChange.java

```diff
@@ -28,75 +27,57 @@
 
 	public UpdateFBInstanceChange(final FBNetworkElement instance, final TypeEntry typeEntry) {
-		this.instance = instance;
+		super(("Update FB instances - " + FBNetworkHelper.getFullHierarchicalName(instance)),
+				EcoreUtil.getURI(instance), FBNetworkElement.class);
 		this.typeEntry = typeEntry;
-	}
 
-	public UpdateFBInstanceChange(final FBNetworkElement instance) {
-		this(instance, null);
 	}
 
 	@Override
-	public String getName() {
-		return "Update FB instances - " + FBNetworkHelper.getFullHierarchicalName(instance); //$NON-NLS-1$
+	public void initializeValidationData(final FBNetworkElement element, final IProgressMonitor pm) {
+		// no additional ValidationData needed
 	}
 
 	@Override
-	public void initializeValidationData(final IProgressMonitor pm) {
-		// can't think of a way to validate an update
-	}
-
-	private Command getCommand(final FBNetworkElement instance) {
-		if (instance instanceof final SubApp subApp && !subApp.isTyped()) {
-			return new UpdateUntypedSubAppInterfaceCommand(instance, (DataTypeEntry) typeEntry);
-		}
+	public RefactoringStatus isValid(final FBNetworkElement element, final IProgressMonitor pm)
+			throws CoreException, OperationCanceledException {
 
-		if (instance instanceof final StructManipulator muxer) {
-			final LibraryElement structuredType = typeEntry.getTypeEditable();
-			Assert.isTrue(structuredType instanceof StructuredType);
-			return new ChangeStructCommand(muxer, (StructuredType) structuredType);
+		final RefactoringStatus status = new RefactoringStatus();
+		if (element.eContainer() == null) {
+			status.addError(element.getQualifiedName() + " eContainer is null"); //$NON-NLS-1$
 		}
-		if (instance.eContainer() instanceof BaseFBType && instance instanceof final FB fb) {
-			return new UpdateInternalFBCommand(fb, typeEntry);
-		}
-
-		return new UpdateFBTypeCommand(instance, typeEntry);
+		return status;
 	}
 
 	@Override
-	public RefactoringStatus isValid(final IProgressMonitor pm) throws CoreException, OperationCanceledException {
-		return new RefactoringStatus();
-	}
+	protected Command createCommand(final FBNetworkElement element) {
 
-	@Override
-	public Change perform(final IProgressMonitor pm) throws CoreException {
-		final var cmd = getCommand(instance);
-		AbstractLiveSearchContext.executeAndSave(cmd, instance, pm);
-		if (cmd instanceof final AbstractUpdateFBNElementCommand ab) {
-			return new UpdateFBInstanceChange(ab.getNewElement(), typeEntry);
+		if (element instanceof final SubApp subApp && !subApp.isTyped()
+				&& typeEntry instanceof final DataTypeEntry dtEntry) {
+			return new UpdateUntypedSubAppInterfaceCommand(element, dtEntry);
 		}
-		return new UpdateFBInstanceChange(instance, typeEntry);
-	}
 
-	@Override
-	public Object getModifiedElement() {
-		return null;
+		if (element instanceof final StructManipulator muxer
+				&& typeEntry.getType() instanceof final StructuredType structuredType) {
+			return new ChangeStructCommand(muxer, structuredType);
+		}
```

Comment:
I think it could also be a configureable function block (e.g., F_MOVE) then you have to perform the update command on this. As structManipulator is a child of configureableFB the check has to be performed afterwards.

Reply:
Should we make a generic command for that ?
I think it is very error-prone when we have to do that on various places.
I think of something like making the constructor private and make a factory method which creates the correct command.
What do you think ?